### PR TITLE
 Add solutions for the 5th assignment

### PR DESCRIPTION
**Pull request by nfadeeva**

### CODE REVIEW

File: nfadeeva/assignment5/find_alphabet.py [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+"""
+
+from typing import List
+from collections import defaultdict
+from collections import deque
+
+
+class Graph:
+    def __init__(self):
+        self.adj = defaultdict(list)
```
Comment (aliskin) [inline nfadeeva/assignment5/find_alphabet.py]: avoid using abbreviations: https://github.com/google/styleguide/blob/gh-pages/pyguide.md#s3.16-naming

File: nfadeeva/assignment5/find_alphabet.py [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    :param dictionary: List of words
+    :returns: List of characters or None if the given dictionary is inconsistent
+    """
+
+    # create graph from the given dictionary
+    graph = create_graph(dictionary)
+    # return the characters in topological order as the requirement alphabet
+    if graph.vertices:
+        return topological_sort(graph, list(graph.vertices)[0])
+    else:
```
Comment (aliskin) [inline nfadeeva/assignment5/find_alphabet.py]: You don'n need else here

File: nfadeeva/assignment5/find_alphabet.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+    graph_list = graph.adj
+    for neighbour in graph_list[v]:
+
+        # neighbour hasn't been visited yet
+        if not visited[neighbour]:
+            dfs(graph, neighbour, visited, order)
+
+        # cycle was found => no topological sort => dictionary is inconsistent
+        elif visited[neighbour] == 1:
+            raise BaseException("The given dictionary is inconsistent")
```
Comment (aliskin) [inline nfadeeva/assignment5/find_alphabet.py]: Actually, your dfs function does not know about dictionary, it operates with a graph. so I'd suggest to write that graph has a cycle.

File: nfadeeva/assignment5/tests.py [hunk 4]
```diff
@@ -0,0 +1,20 @@
+from assignment5.find_alphabet import find_alphabet
+import unittest
```
Comment (aliskin) [inline nfadeeva/assignment5/tests.py]: import standard libraries before other packages.

File: nfadeeva/assignment5/find_alphabet.py (hunk 1 above)
Comment (nfadeeva) [reply nfadeeva/assignment5/find_alphabet.py]: @aliskin thank you for your comments, fixed all places, check it)

File: group/assignment5/find_alphabet.py [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    or the list of vertices in topological sort order
+    """
+    visited = dict.fromkeys(graph.vertices, 0)
+    order = deque()
+
+    # go through all vertices and run the updated dfs
+    # if there is a cycle in the given graph => raise Error and return None
+    for v in graph.vertices:
+        if not visited[v]:
+            if not dfs(graph, v, visited, order):
```
Comment (lizaku) [inline group/assignment5/find_alphabet.py:88]: Can you combine these two conditions? I think it'll look better

**Review by lizaku — APPROVED** [review]
Approved!

**Pull request merged**